src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/VisualStudioRuleSetManager.RuleSetFile.cs

```diff
@@ -154,8 +186,11 @@ private void IncludeUpdated(object sender, EventArgs e)
 
             private void IncludeUpdateCore()
             {
-                _ruleSetManager.StopTrackingRuleSetFile(this);
-                UnsubscribeFromFileTrackers();
+                // It's critical that Dispose() is called first prior to raising the event -- this way any callers who call the RuleSetManager
+                // asking for the new file are guaranteed to get the new snapshot first. It's also strange in that we are calling
+                // Dispose even though we're held by a ReferenceCountedDisposable, but that's fine -- Dispose() is (and should be)
+                // idempotent.
+                Dispose();
```

Comment:
❗️ This breaks an invariant of `ReferenceCountedDisposable<T>`: If you obtain a non-null reference using `TryGetReference` and have not called `Dispose()`, then the target object is not disposed. This is an essential invariant for reasoning about the type at scale.